src/_zkapauthorizer/sql.py

```diff
@@ -5,15 +5,98 @@
 to support testing the replication/recovery system.
 """
 
+from __future__ import annotations
+
 from enum import Enum, auto
-from sqlite3 import Cursor
-from typing import Union
+from sqlite3 import Connection as _SQLite3Connection
+from typing import Any, ContextManager, Iterable, Optional, Protocol, Union
 
-from attrs import define
+from attrs import frozen
 
 SQLType = Union[int, float, str, bytes, None]
 
 
+class AbstractCursor(Protocol):
+    """
+    A SQLite3 database cursor.
+    """
+
+    def execute(self, statement: str, args: Iterable[Any]) -> AbstractCursor:
+        ...
+
+    def close(self) -> None:
+        ...
+
+    def fetchall(self) -> list[Any]:
+        ...
+
+
+class AbstractConnection(Protocol):
+    """
+    A SQLite3 database connection.
+    """
+
+    def iterdump(self) -> Iterable[str]:
+        ...
+
+    def cursor(self, cursorClass: Optional[type] = None) -> AbstractCursor:
+        ...
+
+    def __enter__(self) -> ContextManager:
+        ...
+
+    def __exit__(
+        self,
+        exc_type: Optional[type],
+        exc_value: Optional[BaseException],
+        exc_tb: Optional[Any],
+    ) -> None:
+        ...
+
+
+Connection = AbstractConnection
+Cursor = AbstractCursor
+
+
+class UnboundConnect(Protocol):
```

Comment:
It turns out we needed annotations for both "connect that accepts a path" and "connect that is already partially applied to a path" so there are two of these now.